### PR TITLE
Add APIs for CloudMigrate

### PR DESCRIPTION
**Pull request by disrani-px**

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -417,6 +421,62 @@ type CloudBackupSchedEnumerateResponse struct {
...
+	Status          MigrateStatus
+	LastUpdate      time.Time
+	LastSuccess     time.Time
+}
+
+type CloudMigrateStartRequest struct {
+	Operation Operation
+	ClusterId string
+	// Depending on Operation type this can be VolumeID or VolumeGroupID
+	TargetId string
```
Comment (talakad) [inline api/api.go]: for migrating volume group, should we allow labels as well ( along with group iD)? Since group snapshot api allows both groupID and labels.

File: api/api.go [hunk 2]
```diff
@@ -417,6 +421,62 @@ type CloudBackupSchedEnumerateResponse struct {
...
+}
+
```
Comment (talakad) [inline api/api.go]: might be a good idea to add migrateStartResponse structure that returns a unique ID? this unique ID can then be used to get status and also for cancelling the request?  With volume group migration( using volume labels), both status and cancel requests becomes easier with unique id.

Comment (disrani-px) [reply api/api.go]: The way it is right now, you would need to give it the same targetID for starting and canceling the operation. Migrating the group would take in labels, and you would pass in labels to both. We can update these to return task IDs once we have corresponding task IDs for cloudsnaps.

**Comment by lpabon** [comment]
If these APIs are to be used by an external (non-px) caller, then we should add support also in the SDK.

**Comment by disrani-px** [comment]
This isn't required in the SDK

**Comment by lpabon** [comment]
Ok, I'll review this tonight

File: api/api.go [hunk 3]
```diff
@@ -105,11 +105,15 @@ const (
...
-	OsdSnapshotPath = "osd-snapshot"
-	OsdCredsPath    = "osd-creds"
-	OsdBackupPath   = "osd-backup"
-	TimeLayout      = "Jan 2 15:04:05 UTC 2006"
+	OsdVolumePath        = "osd-volumes"
+	OsdSnapshotPath      = "osd-snapshot"
+	OsdCredsPath         = "osd-creds"
+	OsdBackupPath        = "osd-backup"
+	OsdMigratePath       = "osd-migrate"
+	OsdMigrateStartPath  = OsdMigratePath + "/start"
```
Comment (lpabon) [inline api/api.go]: Why have these defined here? They should be defined in the routes instead as the others are.

File: api/api.go [hunk 4]
```diff
@@ -437,6 +441,62 @@ type CloudBackupSchedEnumerateResponse struct {
 	Schedules map[string]CloudBackupScheduleInfo
 }
 
+type Operation string
+
+const (
```
Comment (lpabon) [inline api/api.go]: Define all of these in api.proto to make it easier when we add this support to the SDK. They will need to be enums. Check [SDK Requirements](https://github.com/libopenstorage/openstorage/blob/master/docs/dev-sdk.md#sdk-requirements) for enums requirements

File: api/api.go [hunk 5]
```diff
@@ -437,6 +441,62 @@ type CloudBackupSchedEnumerateResponse struct {
...
+	LocalVolumeName string
+	RemoteVolumeId  string
+	CloudSnapId     string
+	CurrentStage    MigrateStage
+	Status          MigrateStatus
+	LastUpdate      time.Time
+	LastSuccess     time.Time
+}
+
+type CloudMigrateStartRequest struct {
```
Comment (lpabon) [inline api/api.go]: Move structs as messages in the api.proto file.

File: api/api.go [hunk 6]
```diff
@@ -437,6 +441,62 @@ type CloudBackupSchedEnumerateResponse struct {
...
+	Operation Operation
+	ClusterId string
+	// Depending on Operation type this can be VolumeID or VolumeGroupID
+	TargetId string
```
Comment (lpabon) [inline api/api.go]: Change this to be volume_target_id

File: api/api.go [hunk 7]
```diff
@@ -437,6 +441,62 @@ type CloudBackupSchedEnumerateResponse struct {
...
+	CloudSnapId     string
+	CurrentStage    MigrateStage
+	Status          MigrateStatus
+	LastUpdate      time.Time
+	LastSuccess     time.Time
+}
+
+type CloudMigrateStartRequest struct {
+	Operation Operation
+	ClusterId string
```
Comment (lpabon) [inline api/api.go]: Is this the destination cluster id or the source?

File: api/api.go [hunk 8]
```diff
@@ -437,6 +441,62 @@ type CloudBackupSchedEnumerateResponse struct {
...
+	TargetId string
+}
+
+type CloudMigrateCancelRequest struct {
+	Operation Operation
+	// Depending on Operation type this can be VolumeID or VolumeGroupID
+	TargetId string
+}
+
+type CloudMigrateStatusResponse struct {
```
Comment (lpabon) [inline api/api.go]: What is the key in this map? Please document.

File: api/client/volume/client.go [hunk 9]
```diff
@@ -441,7 +439,7 @@ func (v *volumeClient) Unquiesce(volumeID string) error {
 // CredsEnumerate enumerates configured credentials in the cluster
 func (v *volumeClient) CredsEnumerate() (map[string]interface{}, error) {
 	creds := make(map[string]interface{}, 0)
-	err := v.c.Get().Resource(credsPath).Do().Unmarshal(&creds)
+	err := v.c.Get().Resource(api.OsdCredsPath).Do().Unmarshal(&creds)
```
Comment (lpabon) [inline api/client/volume/client.go]: Why is this change needed? These are the types of things that would be good to document to the reviewer on the PR.

File: volume/drivers/fake/fake.go [hunk 10]
```diff
@@ -51,6 +51,7 @@ type driver struct {
 	volume.QuiesceDriver
 	volume.CredsDriver
 	volume.CloudBackupDriver
+	volume.CloudMigrateDriver
```
Comment (lpabon) [inline volume/drivers/fake/fake.go]: Implement an in-memory implementation of this in the fake driver.

File: api/api.go (hunk 4 above)
Comment (disrani-px) [reply api/api.go]: Don't want to define them as enum, want to use strings like the cloudbackup API types. Can't define enums in the proto file so keeping them here.

File: api/api.go (hunk 5 above)
Comment (disrani-px) [reply api/api.go]: They depend on the typed consts above, so can't move them there

File: api/api.go (hunk 6 above)
Comment (disrani-px) [reply api/api.go]: This can be volume of group id. That's why it is called target id

File: api/client/volume/client.go (hunk 9 above)
Comment (disrani-px) [reply api/client/volume/client.go]: "OsdCredsPath" is already defined in api.go, no reason to define them locally here too.

File: api/api.go (hunk 3 above)
Comment (disrani-px) [reply api/api.go]: The way it is defined right now for the other paths is incorrect. The server and client either redefine them or hard code the paths. The server and client need to refer to them from a common place, which is this file

File: volume/drivers/fake/fake.go (hunk 10 above)
Comment (disrani-px) [reply volume/drivers/fake/fake.go]: I don't see any value in adding that here. The implementation should test it.

**Comment by disrani-px** [comment]
Not sure what more info you want, it's just the APIs that are defined in the PR.

I'll add the Unit tests, but they are kind of useless and don't really provide any value or catch any issues.

File: api/api.go (hunk 3 above)
Comment (lpabon) [reply api/api.go]: Oh I see. Great point. In other words, there should be a common location for both to share. Please create an issue to fix it so that this style can be applied to all.

File: api/api.go (hunk 4 above)
Comment (lpabon) [reply api/api.go]: I understand what you mean, but these values must be defined in api.proto and not here for when the SDK does call this.

File: api/api.go (hunk 5 above)
Comment (lpabon) [reply api/api.go]: They need to move to have them used by the SDK in the future.

File: api/api.go (hunk 6 above)
Comment (lpabon) [reply api/api.go]: As the author of the code this makes sense, but as a reader it is difficult to get that information from `TargetId`. As a reader, it would make it easier to have variables be self-documenting.

File: api/client/volume/client.go (hunk 9 above)
Comment (lpabon) [reply api/client/volume/client.go]: Cool

File: volume/drivers/fake/fake.go (hunk 10 above)
Comment (lpabon) [reply volume/drivers/fake/fake.go]: Need the implementation in fake driver so that when this is added to the SDK, the user of the SDK can use the mock-sdk-server container and test their code. If this is not added at the time of creation, it will must likely never get added by the original author.

**Comment by lpabon** [comment]
> I'll add the Unit tests, but they are kind of useless and don't really provide any value or catch any issues.

This is absolutely incorrect. Unit tests provide an amazing method to make sure that the code you have added works now and most importantly in future changes by others. That is where unit tests provide the most quality assurance.

**Comment by disrani-px** [comment]
I agree with unit tests, but the interface tests that are added for these REST APIs don't really test anything.

File: api/api.go (hunk 4 above)
Comment (disrani-px) [reply api/api.go]: @jvinod mentioned we can still get the string of the enum from the protobuf defintion so moved it to the proto file

File: api/api.go (hunk 7 above)
Comment (disrani-px) [reply api/api.go]: Migration is always to a remote cluster, we don't ever need to pass in the local cluster ID for anything

File: api/api.go (hunk 6 above)
Comment (disrani-px) [reply api/api.go]: There is a comment to clarify it. volume_target_id doesn't make sense since it can also be a group id.

File: api/server/sdk/api/api.swagger.json [hunk 11]
```diff
@@ -17,7 +17,7 @@
   "paths": {
     "/v1/cloudbackup": {
       "get": {
-        "summary": "Enumerate",
+        "summary": "Erases an alert for a given resource",
```
Comment (disrani-px) [inline api/server/sdk/api/api.swagger.json]: `make docs` seems to have messed this up.

Comment (disrani-px) [reply api/server/sdk/api/api.swagger.json]: `make proto` made these changes, I reverted them. But `make proto` still produces this wrong output.

File: api/api.proto [hunk 12]
```diff
@@ -1585,3 +1585,81 @@ message SdkCloudBackupStateChangeRequest {
...
+    // Migrate all volumes in the cluster
+    MigrateCluster = 0;
+    // Migrate a single volume
+    MigrateVolume = 1;
+    // Migrate a group of volumes
+    MigrateVolumeGroup = 2;
+}
+
+enum CloudMigrateStage {
+    Backup = 0;
```
Comment (lpabon) [inline api/api.proto]: All enums must be unique across the entire file (according to protoc error messages). Please prefix with the enum name.

File: api/api.proto [hunk 13]
```diff
@@ -1585,3 +1585,81 @@ message SdkCloudBackupStateChangeRequest {
 
 message SdkCloudBackupStateChangeResponse {
 }
+
+enum CloudMigrateOperationType {
+    // Migrate all volumes in the cluster
+    MigrateCluster = 0;
```
Comment (lpabon) [inline api/api.proto]: Highly suggest that 0 is left as an unknown value on all enums. That way you can check for errors if it was not defined.

File: volume/drivers/fake/fake.go (hunk 10 above)
Comment (lpabon) [reply volume/drivers/fake/fake.go]: This still needs to be done.

File: api/api.proto (hunk 12 above)
Comment (disrani-px) [reply api/api.proto]: Scoped them inside a message which is easier to read then prefixing with the enum name

File: api/api.proto (hunk 13 above)
Comment (disrani-px) [reply api/api.proto]: Added invalid states
